packages/remix-server-runtime/server.ts

```diff
@@ -222,135 +223,53 @@ async function handleDocumentRequestRR(
-      // Internal Router errors will throw an ErrorResponse with the actual
-      // error instance, while user-thrown ErrorResponses will not have the
-      // instance.  We also exclude 404s so we can handle them as CatchBoundary
-      // errors so the user has a singular location for 404 UI
-      if (error.internal && error.error && error.status !== 404) {
-        if (hasErrorBoundary) {
-          appState.loaderBoundaryRouteId = id;
-        }
-        appState.trackBoundaries = false;
-        appState.error = await serializeError(error.error);
-
-        if (
-          error.status === 405 &&
-          error.error.message.includes("Invalid request method")
-        ) {
-          // Note: Emptying this out ensures our response matches the Remix-flow
-          // exactly, but functionally both end up using the root error boundary
-          // if it exists.  Might be able to clean this up eventually.
-          context.matches = [];
-        }
-        break;
-      }
-
-      // ErrorResponse's without an error were thrown by the user action/loader
-      if (hasCatchBoundary) {
-        appState.catchBoundaryRouteId = id;
-      }
-      appState.trackCatchBoundaries = false;
-      appState.catch = {
-        // Order is important for response matching assertions!
-        data:
-          error.error && error.status === 404
-            ? error.error.message
-            : error.data,
-        status: error.status,
-        statusText: error.statusText,
-      };
-      break;
-    } else {
-      if (hasErrorBoundary) {
-        appState.loaderBoundaryRouteId = id;
-      }
-      appState.trackBoundaries = false;
-      appState.error = await serializeError(error);
-      break;
-    }
-  }
-
-  let renderableMatches = getRenderableMatches(
-    context.matches as unknown as RouteMatch<ServerRoute>[],
-    appState
-  );
-
-  if (!renderableMatches) {
-    renderableMatches = [];
-
-    let root = staticHandler.dataRoutes[0] as ServerRoute;
-    if (root?.module?.CatchBoundary) {
-      appState.catchBoundaryRouteId = "root";
-      renderableMatches.push({
-        params: {},
-        pathname: "",
-        route: staticHandler.dataRoutes[0] as ServerRoute,
-      });
-    }
-  }
-
-  let headers = getDocumentHeadersRR(build, context, renderableMatches);
-
-  let serverHandoff: Pick<
-    EntryContext,
-    "actionData" | "appState" | "matches" | "routeData" | "future"
-  > = {
-    actionData: context.actionData || undefined,
-    appState,
-    matches: createEntryMatches(renderableMatches, build.assets.routes),
-    routeData: context.loaderData || {},
-    future: build.future,
-  };
+  let headers = getDocumentHeadersRR(build, context);
 
   let entryContext: EntryContext = {
-    ...serverHandoff,
     manifest: build.assets,
     routeModules: createEntryRouteModules(build.routes),
-    serverHandoffString: createServerHandoffString(serverHandoff),
+    staticHandlerContext: context,
```

Comment:
No more `appState` - just work directly with the `@remix-run/router` `staticHandlerContext` now